keps/sig-instrumentation/20200415-cardinality-enforcement.md

```diff
@@ -0,0 +1,272 @@
+2. metrics cardinality issues exist latently; in other words, they're all over the place in Kubernetes already (so even if we could prevent 100% of these occurrence from **now on**, that wouldn't guarantee Kubernetes is free of these classes of issues).
+
+Instead, the proposed solution is we will be able to *dynamically configure* an allowlist of label values for a metric. By *dynamically configure*, we mean configure an allowlist *at runtime* rather than during the build/compile step. And by *at runtime*, we mean, more specifically, during the boot sequence for a Kubernetes component (and we mean daemons here, not CLI tools like kubectl).
+
+Brief aside: a Kubernetes component (which is a daemon) is an executable, which can be launched from the command line manually if desired. Components take a number of start-up configuration flags, which are passed into the component to modify execution paths (if curious, check out the [large amount of flags we have on the kubelet](https://kubernetes.io/docs/reference/command-line-tools-reference/kubelet/)). It is also possible to read configuration data from files (like yaml format) during the component boot sequence. This KEP does not have an opinion on the specific mechanism used to load config data into a Kubernetes binary during the boot sequence. What we *actually* care about, is just the fact that it is possible.
+
+Our design is thus config-ingestion agnostic.
+
+Our design is also based on the premise that metrics can be uniquely identified (i.e. by their metric descriptor). Fortunately for us, this is actually a built in constraint of prometheus clients (which is how we instrument the Kubernetes components). This metric ID is resolvable to a unique string (this is pretty evident when looking at a raw prometheus client endpoint).
+
+We want to provide a data structure which basically maps a unique metric ID and one of it's labels to a bounded set of values.
+
+Purely for illustration, take the following metric:
+
+```json
+[
+	{
+		"metric-id": "some_metric"
+	}
+]
+
+```
+
+We want to be able to express this metric and target a label:
+
+
+```json
+[
+	{
+		"metric-id": "some_metric",
+		"label": "label_too_many_values",
+	}
+]
+
+```
+
+And we want to express a set of expected values:
+
+
+```json
+[
+	{
+		"metric-id": "some_metric",
+		"label": "label_too_many_values",
+		"labelValueAllowlist": [
+			"1",
+			"2",
+			"3"
+		]
+	}
+]
+
+```
+
+Since we already have an interception layer built into our Kubernetes monitoring stack (from the metrics stability effort), we can leverage existing wrappers to provide a global entrypoint for intercepting and enforcing rules for individual metrics. 
+
+As a result metric data will not be invalid, just data will be more coarse when the metric type is a counter. In the case of a gauge type metric the data will be invalid but the number of series would be bound.
```

Comment:
what about histograms? I'd be ok if histograms were a non-goal, but that would need to be explicit, they are a composite metric made up of counters though so we should be able to apply similar logic

Reply:
Added a note.